docs/proposals/scheduling/20220629-schedule-device-in-card-level.md

```diff
@@ -0,0 +1,268 @@
+type NodeDevice struct {
+    DeviceTotal map[DeviceType]*DeviceResource
+    DeviceFree  map[DeviceType]*DeviceResource
+    DeviceUsed  map[DeviceType]*DeviceResource
+    AllocateSet map[string]*PodInfo
+}
+
+type NodeDeviceCache struct {
+    nodeDeviceMap map[string]*NodeDevice
+}
+
+type NodeDevicePlugin struct {
+    frameworkHandler     framework.Handle
+    deviceClient         deviceClient.Interface
+    deviceLister         devicelister.DeviceLister
+    nodeLister           listerv1.NodeLister
+    nodeDeviceCache      *NodeDeviceCache
+}
+
+type NodeDevicePlugin interface{
+    MonotorLoop()
+    Filter(ctx context.Context, state *CycleState, pod *v1.Pod, filteredNodeStatusMap NodeToStatusMap) (*PostFilterResult, *Status)
+    Reserve(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status
+    UnReserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string)
+    PreBind(context.Context, *corev1.Pod, string)
+}
+```
+
+in MonitorLoop, we will update NodeDeviceCache by deviceLister and nodeLister.
+
+in Filter, we will make-up each device request by concrete node(the gpu-mem example), and try compare each device free 
+resource and pod device request.
+
+in Reserve\UnReserve, we will update nodeDeviceCache's used\free resource and allocateSet.
+
+in PreBind, we will write DeviceAllocations to pod's annotation。
+
+#### device reporter
+we need create a new device reporter on node to update a new device crd to report each device card resource and health status
+
+```go
+type DeviceType string
+
+const (
+    GPU  DeviceType = "gpu-device"
+    FPGA DeviceType = "fpga-device"
+    RDMA DeviceType = "rdma-device"
+)
+
+type DeviceList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items  []Device `json:"items"`
+}
+
+type Device struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    NodeName     string        `json:"nodeName"`
+    DeviceType   DeviceType    `json:"deviceType"`
+    Spec         DeviceSpec    `json:"spec,omitempty"`
+}
+
+type DeviceSpec struct {
+    DeviceInfos     []DeviceInfo    `json:"deviceInfos"`
+    deviceTopology  DeviceTopology  `json:"deviceTopology"`
+}
+
+type DeviceInfo struct {
+    ID        string                       `json:"id"`       // Device UUID
+    Resources map[string]resource.Quantity `json:"resource"` // map of sub-resources
+    Health    bool                         `json:"health"`
+    Minor     int32                        `json:"minor"` // index starting from 0
+}
+
+type DeviceTopology struct {
+    Links [][]DeviceLink `json:"links,omitempty"`
+}
+type DeviceLink struct {
+    Type      string      `json:"LinkType,omitempty"`
+    Bandwidth int64       `json:"bandwidth,omitempty"`
+}
+
+this version we only provide gpu device reporter. it will execution nvidia-smi commands to get each minor resource just
```

Comment:
Is the `gpu device reporter` a daemonset? Or implement the reporter in koordlet? Need to clarify this detail.